pkg/controller/database/postgresqlserver/managed.go

```diff
@@ -0,0 +1,212 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package postgresqlserver
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/postgresql/mgmt/postgresql"
+	"github.com/negz/crossplane/pkg/util"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	runtimev1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane-runtime/pkg/meta"
+	"github.com/crossplaneio/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplaneio/stack-azure/apis/database/v1alpha2"
+	azurev1alpha2 "github.com/crossplaneio/stack-azure/apis/v1alpha2"
+	azure "github.com/crossplaneio/stack-azure/pkg/clients"
+)
+
+const passwordDataLen = 20
+
+// Error strings.
+const (
+	errNewClient                 = "cannot create new PostgresqlServer client"
+	errGetProvider               = "cannot get Azure provider"
+	errGetProviderSecret         = "cannot get Azure provider Secret"
+	errGenPassword               = "cannot generate admin password"
+	errNotPostgresqlServer       = "managed resource is not a PostgresqlServer"
+	errCreatePostgresqlServer    = "cannot create PostgresqlServer"
+	errGetPostgresqlServer       = "cannot get PostgresqlServer"
+	errDeletePostgresqlServer    = "cannot delete PostgresqlServer"
+	errCheckPostgresqlServerName = "cannot check PostgresqlServer name availability"
+)
+
+// Controller is responsible for adding the PostgresqlServer controller and its
+// corresponding reconciler to the manager with any runtime configuration.
+type Controller struct{}
+
+// SetupWithManager creates a new PostgresqlServer Controller and adds it to the
+// Manager with default RBAC. The Manager will set fields on the Controller and
+// start it when the Manager is Started.
+func (c *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1alpha2.PostgresqlServerGroupVersionKind),
+		resource.WithExternalConnecter(&connecter{client: mgr.GetClient(), newClientFn: newClient}))
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1alpha2.PostgresqlServerKind, v1alpha2.Group))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha2.PostgresqlServer{}).
+		Complete(r)
+}
+
+func newClient(credentials []byte) (azure.PostgreSQLServerAPI, error) {
+	ac, err := azure.NewClient(credentials)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create Azure client")
+	}
+	pc, err := azure.NewPostgreSQLServerClient(ac)
```

Comment:
In regards to my earlier comment, maybe it is more informative to have these two separate client creation steps? I am not sure.. I don't know if there would ever be a case in which we could create the general client but then fail to create the service specific one.

Reply:
Yeah, mostly the service specific client creation ones are doing very little. I think I'll change this to match your preference.

Reply:
I agree with the direction, but ended up punting on this in the interest of time.